[clang][analyzer] Fix argument invalidations in StreamChecker.

balazske opened the pull request:
Specific arguments passed to stream handling functions are changed by the function, this means these should be invalidated ("escaped") by the analyzer. This change adds the argument invalidation (in specific cases) to the checker.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -544,6 +545,21 @@ const ExplodedNode *StreamChecker::getAcquisitionSite(const ExplodedNode *N,
 }
 
+static ProgramStateRef
+escapeArgs(ProgramStateRef State, CheckerContext &C, const CallEvent &Call,
+           const SmallVector<unsigned int> &EscapingArgs) {
```
```suggestion
escapeArgs(ProgramStateRef State, CheckerContext &C, const CallEvent &Call,
           ArrayRef<unsigned int> EscapingArgs) {
```
Views should be preferred for input parameters.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -544,6 +545,21 @@ const ExplodedNode *StreamChecker::getAcquisitionSite(const ExplodedNode *N,
+  EscapingVals.reserve(EscapingArgs.size());
+  for (auto EscArgIdx : EscapingArgs)
+    EscapingVals.push_back(Call.getArgSVal(EscArgIdx));
+  State = State->invalidateRegions(EscapingVals, CE, C.blockCount(),
+                                   C.getLocationContext(),
+                                   /*CausesPointerEscape=*/false);
```
```suggestion
  State = State->invalidateRegions(EscapingVals, CE, C.blockCount(),
                                   C.getLocationContext(),
                                   /*CausesPointerEscape=*/false, /*InvalidatedSymbols=*/nullptr, &Call);
```

I can't recall now what difference it makes to pass the `Call` to this API, but given that we have one, why not?

balazske replied:
If this change is made, a error shows up in **stream-errno.c** line 190 (unexpected warning: FALSE).
```
  int Ret = fgetpos(F, &Pos);
  if (Ret)
    clang_analyzer_eval(errno != 0); // expected-warning{{TRUE}}
  else
    clang_analyzer_eval(errno == 0); // expected-warning{{TRUE}} we get FALSE too
```
Probably the `errno` is invalidated somehow, this is likely to be caused by the `Call` argument. Without these additional arguments to `invalidateRegions` the test passes.

steakhal replied:
Yup, have a look at `RegionStoreManager::invalidateRegions`. Depending on the nature of the call, invalidates system or every global variables.
We don't really care about it, so let's just not pass the `Call` here.

steakhal commented on clang/test/Analysis/stream-invalidate.c:
```diff
@@ -0,0 +1,133 @@
+  // this check applies to succes and failure
+  clang_analyzer_dump(Buf[0]); // expected-warning {{conj_$}} Should not preserve the previous value, thus should not be 10.
+  clang_analyzer_dump(Buf[2]); // expected-warning {{conj_$}}
+  if (feof(F)) {
```
In this, and the in the rest of the test, could we have a branch for testing if the api-call encountered an error?
E.g. when `ferror(F)==true`. On that path, we should still invalidate the buffer.

balazske replied:
At `ferror(F)` the next stream call should produce a "file position indeterminate" warning (this is tested in another file) that stops the analysis, buffer invalidation is not needed to be tested.

steakhal commented on clang/test/Analysis/stream-invalidate.c:
```diff
@@ -0,0 +1,133 @@
+
+  char Buf[3] = {10, 10, 10};
+  fwrite(Buf, 1, 3, F);
+  // this check applies to succes and failure
```
```suggestion
  // This check applies to success and failure.
```
As per llvm style, capitalize and punctuate comments.
`sed "s/this check applies to succes and failure/This check applies to success and failure."`
There was also a typo `sed "s/succes/success/g"` this file.

steakhal commented on clang/test/Analysis/stream-invalidate.c:
```diff
@@ -0,0 +1,133 @@
+  int Ret = fscanf(F, "%d %u", &a, &b);
+  if (Ret >= 0) {
+    // FIXME: return value
+    clang_analyzer_dump(a); // expected-warning {{conj_$}}
```
```suggestion
    clang_analyzer_dump(a); // expected-warning {{conj_$}}
    clang_analyzer_eval(Ret > 2); // expected-warning {{FALSE}} expected-warning {{TRUE}} FIXME: should be only FALSE.
```
Let's take the opportunity for adding an expected (but failing) assumption here. May be fixed in the future.

steakhal commented on clang/test/Analysis/stream-invalidate.c:
```diff
@@ -0,0 +1,133 @@
+  char *P1 = P;
+  size_t Sz1 = Sz;
+  ssize_t Ret = getdelim(&P, &Sz, '\t', F);
+  clang_analyzer_eval(P == P1); // expected-warning {{FALSE}} \
+                                // expected-warning {{TRUE}}
+  clang_analyzer_eval(Sz == Sz1); // expected-warning {{FALSE}} \
+                                  // expected-warning {{TRUE}}
```
I don't think this is too useful this way. We can't distinguish the failure and success branches.
Lets have different branches depending on `Ret`, to make it clear.

steakhal reviewed (commented):
I like what you do in this patch.
I only have a couple nits. That's it.

Tomorrow, I'll check if there are any other APIs that we should test; but seems complete at first glance.

steakhal reviewed (approved):
Thanks for resolving my comments.
FYI if I forget about a PR (that I promise to come back on the next day) - feel free to ping it or explicitly push the "request review" button.

Wait for my collage to also have a look, as I believe he might be in context to review this change. @alejandro-alvarez-sonarsource

alejandro-alvarez-sonarsource commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -763,6 +779,11 @@ void StreamChecker::evalFreadFwrite(const FnDescription *Desc,
+  // At read, invalidate the buffer in any case of error or success,
+  // except if EOF was already present.
+  if (IsFread && (OldSS->ErrorState != ErrorFEof))
+    State = escapeArgs(State, C, Call, {0});
```
From what I can tell, the lambda `UpdateBufferRegionForFread` is already used to take care of this in a more fine-grained manner. For instance:

```cpp
int buffer[10];
buffer[5] = 42;
if (1 == fread(buffer, sizeof(int), 5, fd)) {
    assert(buffer[5] == 42);
}
```

Before this change, the assertion would pass, since lambda took `nmemb` into account. With this change, the whole buffer is invalidated.

balazske replied:
The current solution invalidates the whole buffer (if I think correctly) because the `SVal` of the buffer is passed to `invalidateRegions`. Is there a solution to invalidate a partial buffer (of specific size)?

alejandro-alvarez-sonarsource replied:
It turns out I was comparing with our downstream copy of v17, where `UpdateBufferRegionForFread` is defined.
So it can be done for the case of arrays. I have double-checked with the original author and he agrees it can be upstreamed. Let me uplift the patch and then I will share it with you.

alejandro-alvarez-sonarsource replied:
Here it is: https://gist.github.com/alejandro-alvarez-sonarsource/48edec4debc8912a6485f989b2a6f0db

I have rebased it on top of 4f12f47550eee85447c9ec37d27a20c6593d3d40 and run check-clang to make sure I didn't break anything.

balazske replied:
This could be a future improvement (after this patch). I am not sure if this works always correct, because it looks like that the `ElementCount` value at `fread` is used but not the "size" argument (size of one element). It is assumed that the specified element size is always the same as one array element, but this can be different.

alejandro-alvarez-sonarsource replied:
Ok, let’s do that, and send this patch as a separate pr. @steakhal LGTM.

balazske committed:
[clang][analyzer] Fix argument invalidations in StreamChecker.

Specific arguments passed to stream handling functions are changed by the function,
this means these should be invalidated ("escaped") by the analyzer.
This change adds the argument invalidation (in specific cases) to the checker.

balazske commented:
I have rebased the branch to latest version of `StreamChecker`, no other changes were made to the patch.

balazske committed:
Add test file
